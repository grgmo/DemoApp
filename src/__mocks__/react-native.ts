import RN from 'react-native';

RN.NativeModules.AppInfo = {
  getConstants: () => ({
    appVersion: '1.0.0',
  }),
};
module.exports = RN;
