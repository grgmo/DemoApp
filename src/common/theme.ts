export default {
  FONT: {
    SMALL: 14,
    MEDIUM: 16,
    LARGE: 18,
    XLARGE: 22,
  },
  SPACE: {
    XSMALL: 5,
    SMALL: 10,
    MEDIUM: 15,
  },
  COLOR: {
    black: 'black',
    lightGrey: 'lightgrey',
    white: 'white',
    whitesmoke: 'whitesmoke',
  },
};
