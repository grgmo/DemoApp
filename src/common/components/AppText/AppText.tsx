import React, {FC, useMemo} from 'react';
import {StyleProp, StyleSheet, Text, TextProps, TextStyle} from 'react-native';
import theme from '../../theme';

export type AppTextType = 'h1' | 'h2' | 'h3' | 'h4';

interface AppTextProps extends TextProps {
  text: string;
  bold?: boolean;
  center?: boolean;
  type?: AppTextType;
}

const AppText: FC<AppTextProps> = ({
  text,
  style,
  bold,
  center,
  type,
  ...props
}) => {
  const newStyle: StyleProp<TextStyle> = useMemo(() => {
    let styleToOverride: StyleProp<TextStyle> = [styles.default];

    if (style) {
      styleToOverride.push(style);
    }

    if (type) {
      styleToOverride.push(styles[type]);
    }

    if (bold) {
      styleToOverride.push(styles.bold);
    }

    if (center) {
      styleToOverride.push(styles.center);
    }

    return styleToOverride;
  }, [style, bold, center, type]);

  return (
    <Text accessibilityRole="text" {...props} style={newStyle}>
      {text}
    </Text>
  );
};

export default AppText;

const styles = StyleSheet.create({
  default: {
    fontSize: theme.FONT.SMALL,
    color: theme.COLOR.black,
  },
  h1: {
    fontSize: theme.FONT.XLARGE,
  },
  h2: {
    fontSize: theme.FONT.LARGE,
  },
  h3: {
    fontSize: theme.FONT.MEDIUM,
  },
  h4: {
    fontSize: theme.FONT.SMALL,
  },
  bold: {
    fontWeight: 'bold',
  },
  center: {
    textAlign: 'center',
  },
});
