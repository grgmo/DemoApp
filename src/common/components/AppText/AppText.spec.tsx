import {render} from '@testing-library/react-native';
import React from 'react';
import theme from '../../theme';
import AppText, {AppTextType} from './AppText';

describe('AppText', () => {
  const mockText = 'text';
  const defaultStyle = {
    fontSize: theme.FONT.SMALL,
    color: theme.COLOR.black,
  };

  it('should render default text correctly', () => {
    const appText = render(<AppText text={mockText} />);

    expect(appText.toJSON()).toMatchSnapshot();
  });

  it.each([
    [
      'h1',
      {
        fontSize: theme.FONT.XLARGE,
      },
    ],
    [
      'h2',
      {
        fontSize: theme.FONT.LARGE,
      },
    ],
    [
      'h3',
      {
        fontSize: theme.FONT.MEDIUM,
      },
    ],
    [
      'h4',
      {
        fontSize: theme.FONT.SMALL,
      },
    ],
  ])('should render h1 text correctly %s %o', (type, expectedStyle) => {
    const {getByRole} = render(
      <AppText type={type as AppTextType} text={mockText} />,
    );

    const {props} = getByRole('text');

    expect(props.style).toEqual([defaultStyle, expectedStyle]);
  });

  it('should render text with center correctly', () => {
    const {getByRole} = render(<AppText center text={mockText} />);

    const {props} = getByRole('text');

    expect(props.style).toEqual([
      defaultStyle,
      {
        textAlign: 'center',
      },
    ]);
  });

  it('should render text with bold correctly', () => {
    const {getByRole} = render(<AppText bold text={mockText} />);

    const {props} = getByRole('text');

    expect(props.style).toEqual([
      defaultStyle,
      {
        fontWeight: 'bold',
      },
    ]);
  });

  it('should render text with h1, bold and center correctly', () => {
    const {getByRole} = render(
      <AppText type="h1" bold center text={mockText} />,
    );

    const {props} = getByRole('text');

    expect(props.style).toEqual([
      defaultStyle,
      {
        fontSize: theme.FONT.XLARGE,
      },
      {
        fontWeight: 'bold',
      },
      {
        textAlign: 'center',
      },
    ]);
  });
});
