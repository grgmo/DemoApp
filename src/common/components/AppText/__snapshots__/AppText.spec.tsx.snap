// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppText should render default text correctly 1`] = `
<Text
  accessibilityRole="text"
  style={
    Array [
      Object {
        "color": "black",
        "fontSize": 14,
      },
    ]
  }
>
  text
</Text>
`;
